### DISCUSSION TITLE
Flapper Drone Interface from external PC

### DISCUSSION THREAD

**Discussion by marios-stam** [bitcraze/discussions#913] [Q&A]
Hello all,

we are trying to interface Flapper Drone from an external PC through the radio. 

After reading [this](https://www.bitcraze.io/2023/04/flapper-drones-and-positioning-systems/) very interesting blog post , if I understood correctly, the only thing we have to do is connect to it through the cflib and command position commands though the high-level commander as a typical Crazyflie (given the fact that there is position feedback from Lighthouse or Loco ).

Since we want to execute our own position controller in the external PC (which is already receiving position feedback from Vicon), is it possible to either send velocity commands (linear x,y,z and yaw-rate) or attitude commands (thrust, roll, pitch, yaw-rate)? 

If yes, would it also be possible to do the same through Crazyswarm since we are aiming to use ROS at the end? 

Thank you in advance!

**Comment by krichardsson** [comment]
Hi Marios!

There are functions in the python lib to send various types of setpoints. Velocity setpoints in the global ref frame is available in [`send_velocity_world_setpoint()`](https://github.com/bitcraze/crazyflie-lib-python/blob/f35b0baf6ff243d618f7a8becf5a69b1ec6c1f56/cflib/crazyflie/commander.py#L113C9-L113C37)

Attitude setpoints is available in [`send_setpoint()`](https://github.com/bitcraze/crazyflie-lib-python/blob/f35b0baf6ff243d618f7a8becf5a69b1ec6c1f56/cflib/crazyflie/commander.py#L70). 

There is also (since a week or two) the [`send_full_state_setpoint()`](https://github.com/bitcraze/crazyflie-lib-python/blob/f35b0baf6ff243d618f7a8becf5a69b1ec6c1f56/cflib/crazyflie/commander.py#L161C9-L161C33) that is very flexible but also more complex. It was added by Crazyswarm together with the Mellinger controller and I think this is the best option for this setpoint, as all controllers will not be able to use all information in this setpoint. I'm not sure if Mellinger and the Flapper is compatible though, @matejkarasek do you have any info on this?

> If yes, would it also be possible to do the same through Crazyswarm since we are aiming to use ROS at the end?

I do not know what kind of set points Crazyswarm is supporting, apart the full state one, you better check with them (or the code).

**Reply by matejkarasek** [reply]
We have not tested Mellinger controller with the Flapper, and I expect some development would be needed to get it working... Only the PID controller is tested and recommended.

We have not tested the Flapper with the crazyswarm yet but possibly @whoenig did already?

**Reply by whoenig** [reply]
Not yet, I am embarrassed to say, but it's very high on our wish list. We are moving to a bigger space by the end of the year and hopefully can test it in there more safely.

**Reply by matejkarasek** [reply]
No worries, good to hear it is in the pipeline... It has been sitting on our list too for quite a while... Good luck with the moving!

@marios-stam In any case, the python libraries should work for you, at least for starters...

**Comment by marios-stam** [comment]
Hello all and thank you very much for your fast answers. 

We managed to make it work with both cflib and crazyswarm!

I used the cflib example [here](https://github.com/bitcraze/crazyflie-lib-python/blob/f35b0baf6ff243d618f7a8becf5a69b1ec6c1f56/examples/motors/ramp.py) just to check that I can interface with the flapper by sending it attitude commands.

After that, we used crazyflie_ros for sending roll, pitch, yaw rate and thrust commands (outdated version of crazyswarm but works for what we want). Finally, we managed to fly it with Vicon as a motion capture system, received data and ran a model predictive controller in the external PC. After some hacking, it worked quite descent and also avoided moving obstacles. With a 2 min. tuning and a wrong model for the platform it worked pretty well!!

I think we need to acknowledge the great work that has been done for these platforms and ecosystems, the integration worked seamlessly and was fun experience!

**Reply by matejkarasek** [reply]
Thanks for sharing, happy to hear the integration went smooth!
